icaruscode/PMT/OpReco/BarycenterMatchProducer_module.cc

```diff
@@ -0,0 +1,548 @@
+  fEvent  = e.id().event();
+  fRun    = e.run();
+  const bool isData = e.isRealData();
+
+  //Fetch trigger info and if MC check whether this event triggered
+  art::Handle<std::vector<raw::Trigger>> triggerHandle;
+  e.getByLabel(fTriggerLabel, triggerHandle);
+  raw::Trigger trigger;
+  double triggerWithinGate = 0.;
+  bool triggeredEvt = false;
+  if ( triggerHandle.isValid() && triggerHandle->size() == 1 ) {
+    trigger = (*triggerHandle).at(0);
+    if ( trigger.TriggerTime() >= 0 ) {
+      triggerWithinGate = trigger.TriggerTime() - trigger.BeamGateTime();
+      triggeredEvt = true;
+    }
+    if ( fVerbose ) std::cout << "Valid trigger product found. Trigger time: " << trigger.TriggerTime() << ", Beam gate time: " << trigger.BeamGateTime() << ", Difference: " << triggerWithinGate  << std::endl;
+  }
+  else if ( fVerbose ) std::cout << "No valid trigger product found for this event!"  << std::endl;
+
+  //Infrastructure for checking allowed time range of a slice
+  microseconds margin(fTimeRangeMargin);
+  detinfo::DetectorTimings const detTimings{ fDetClocks.DataFor(e) };
+  detinfo::DetectorPropertiesData const& detProp { fDetProp.DataFor(e, detTimings.clockData()) };
+  lar::util::TrackTimeInterval const timeIntervals = fTimeIntervalMaker(detProp, detTimings);
+
+  //Initialize new data products
+  auto matchInfoVector = std::make_unique< std::vector<sbn::BarycenterMatch> >();
+  art::PtrMaker< sbn::BarycenterMatch > const makeInfoPtr(e); 
+  auto sliceAssns = std::make_unique< art::Assns<sbn::BarycenterMatch, recob::Slice> >();
+  auto flashAssns = std::make_unique< art::Assns<sbn::BarycenterMatch, recob::OpFlash> >();
+
+  //For InputTag...
+  for ( const std::string inputTag : fInputTags ) {
+    //East-->0, West-->1
+    fCryo = ( inputTag.find("W") != std::string::npos ) ? 1 : 0;
+
+
+/* ~~~~~~~~~~~~~~~~~~~~ Flash Section
+ *
+ * Here we gather the OpFlashes found in this cryostat and their OpHits
+ * We iterate through the flashes to identify a triggering flash
+ */
+
+    //Fetch the flashes and their associated hits, pointer vector needed for generating associations
+    art::Handle<std::vector<recob::OpFlash>> flashHandle;
+    e.getByLabel(fOpFlashLabel + inputTag, flashHandle);
+    art::FindMany<recob::OpHit> fmOpHits(flashHandle, e, fOpFlashLabel + inputTag);
+    std::vector<art::Ptr<recob::OpFlash>> flashVector;
+    art::fill_ptr_vector(flashVector, flashHandle);
+
+    int nFlashes = (*flashHandle).size();
+
+    double triggerFlashCenter = -9999.;
+    double flashTime_Trigger;
+    //For flash...
+    for ( int i = 0; i < nFlashes; i++ ) {
+      const recob::OpFlash &flash = (*flashHandle).at(i);
+
+      //Is this a triggering flash?
+      flashTime_Trigger = flash.Time();
+      if ( !isData ) flashTime_Trigger += -1.*triggerWithinGate;
+      if ( triggeredEvt && abs(flashTime_Trigger - fNominalTrigTime) < fTriggerTolerance ) triggerFlashCenter = flash.ZCenter();
```

Comment:
Thank you, good catch. Nominally my `fTriggerTolerance` is 150ns, so I think it's very unlikely if not impossible to get two of these in the same cryostat. That said, if for whatever reason one wanted to widen that window this would matter more, so it'd be good to account for.